### PR TITLE
Test set generation paradigm

### PR DESCRIPTION
**Pull request by shahules786**
#136 
TODO

- [ ] Add README with flowchart demonstration of paradigm 
- [x] Add quickstart notebook
- [x]  Fix type hinting 
- [x] Reformat methods if needed
- [x] Add progress bar
- [ ] Async workflow

### CODE REVIEW

File: src/ragas/testset/testset_generator.py [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+        results = generate(prompts=[prompt], llm=self.ctitic_llm)
+        return bool(results.generations[0][0].strip().endswith("Yes."))
+
+    def _reasoning_question(self, question: str, context: str):
+        return self._qc_template(REASONING_QUESTION, question, context)
+
+    def _condition_question(self, question: str, context: str):
+        return self._qc_template(CONDITIONAL_QUESTION, question, context)
+
+    def _multicontext_question(self, question: str, context1: str, context2: str):
```
Comment (jjmachan) [inline src/ragas/testset/testset_generator.py:136]: return types are not annotated anywhere actually

File: src/ragas/testset/testset_generator.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+                """Maximum possible number of samples exceeded, 
+                             reduce test_size or add more documents"""
+            )
+
+        available_indices = np.arange(0, len(document_nodes)).tolist()
+        doc_nodeidx = self._generate_doc_node_map(document_nodes)
+        count = 0
+        Testdata_tuple = namedtuple(
+            "Testdata_tuple", ["question", "context", "answer", "question_type"]
+        )
```
Comment (jjmachan) [inline src/ragas/testset/testset_generator.py:215]: we can declare this outside as `TestdataRow` or something so that it can be used as a type

**Review by jjmachan — COMMENTED** [review]
basically 3 things
- type annotation for return types
- more comments: this would really help understand the code better
- there are some small functions that are only used once. should we remove those and add them in-place instead

**Comment by jjmachan** [comment]
also should we keep adding to this PR or merge this and make smaller PRs?

File: src/ragas/testset/testset_generator.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+
+        return embeddings
+
+    def generate(self, documents: t.List[Document], test_size: int):
+        node_parser = SimpleNodeParser.from_defaults(
+            chunk_size=self.chunk_size, chunk_overlap=0, include_metadata=True
+        )
+        document_nodes = node_parser.get_nodes_from_documents(documents=documents)
+
+        if test_size > len(document_nodes):
```
[the remark this answers is not in the document]
Comment (shahules786) [reply src/ragas/testset/testset_generator.py:204]: For this version, I don't recommend generating more than one `seed question` per node. Moving on I have to find a way to ensure diversity even if the algorithms need to frame more questions from a single node.

**Review by jjmachan — APPROVED** [review]

**Pull request merged**